### PR TITLE
add torch.autograd._unsafe_set_version_counter API

### PR DESCRIPTION
**Pull request by bdhirsh**
better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93994
* #92857
* #94195
* #94018
* __->__ #92924

### CODE REVIEW

**Comment by ezyang** [comment]
Instead of manually doing a decrement, wouldn't it make more sense to just allow setting the version counter to some value? This would allow you to do multiple mutating operations and then reset it to some historical state.

**Comment by bdhirsh** [comment]
that seems fair (and imo that also makes it clearer that this is an unsafe API). I'll update

**Review by ezyang — APPROVED** [review]

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -388,6 +388,15 @@ struct C10_API VariableVersion {
     }
   }
 
+  void set_version(int64_t i) {
+    TORCH_CHECK(
+        version_counter_,
+        "Tried to call torch.autograd._unsafe_set_version() on a tensor "
+        "that does not have a version counter. Was it created it in inference mode?");
```
Comment (alanwaketan) [inline c10/core/TensorImpl.h:395]: Nit: Was it created in inference mode?

**Review by alanwaketan — APPROVED** [review]
Thanks, Brian. Let me verify if it fixes the issue.

**Review by albanD — COMMENTED** [review]
I still think that a "decrement" is better than a "set". It give a lot less room for the user to get things wrong. Could you share why you made it a set here?

**Comment by bdhirsh** [comment]
Great! 

> Detailed error:

@alanwaketan just to confirm, are you saying that you're seeing an error? Or everything worked as expected

**Comment by bdhirsh** [comment]
One other thing - @ezyang are you ok with a context manager, like `with torch.autograd._preserve_version_counter(tensor)`, that internally calls this `_unsafe_set_version_counter` API? I was talking to Alban, and he thinks this is slightly less of a footgun: we don't want people doing stuff like this in the future:
```
old_version = t._version
function_that_does_arbitrary_things_to_t(t)
# We might be ignoring legitimate mutations to t that autograd should have been aware of!
torch.autograd._unsafe_set_version_counter(old_version)
```

**Comment by alanwaketan** [comment]
> Great!
> 
> > Detailed error:
> 
> @alanwaketan just to confirm, are you saying that you're seeing an error? Or everything worked as expected

Updated the confusing comment. That was the original error.

**Comment by ezyang** [comment]
I am fine with context manager

File: torch/autograd/grad_mode.py [hunk 2]
```diff
@@ -254,6 +255,35 @@ def __exit__(self, *args) -> None:
     def clone(self):
         return self.__class__(self.mode)
 
+class _unsafe_preserve_version_counter(_DecoratorContextManager):
+    r"""Context-manager that unsafely tells autograd to ignore mutations to a tensor.
```
Comment (albanD) [inline torch/autograd/grad_mode.py:259]: Could you make this more scarry please?
```suggestion
    r"""DO NOT USE THIS UNLESS YOU KNOW EXACTLY WHAT YOU'RE DOING!
    This context manager can lead to arbitrary silent-correctness issues in any other part of your code (even the ones not touched directly by the context manager)!
    
    Context-manager that unsafely tells autograd to ignore mutations to a tensor.
```

**Review by albanD — APPROVED** [review]
Just update to the doc. Sounds good otherwise.

File: torch/autograd/grad_mode.py (hunk 2 above)
Comment (bdhirsh) [reply torch/autograd/grad_mode.py:258]: absolutely 😛

File: torch/autograd/grad_mode.py [hunk 3]
```diff
@@ -254,6 +255,35 @@ def __exit__(self, *args) -> None:
     def clone(self):
         return self.__class__(self.mode)
 
+class _unsafe_preserve_version_counter(_DecoratorContextManager):
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch/autograd/grad_mode.py:257]: Yea, it looks like it would work. I can give it a try as well. Will respond back.

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**